0000-ilp-over-http.md

```diff
@@ -0,0 +1,59 @@
+---
+title: ILP Over HTTP
+draft: 1
+---
+
+# ILP Over HTTP
+> A bilateral communication protocol for server-to-server connections
+
+## Motivation
+
+Scaling Interledger infrastructure to handle large volumes of ILP packets requires horizontally scaling connectors. Using HTTP for bilateral communication enables service providers to leverage standard tools and services for hosting, load balancing, Distributed Denial of Service (DDoS) protection, and monitoring.
+
+## Overview
+
+In an ILP Over HTTP connection, both peers run HTTP servers with accessible HTTPS endpoints. When peering, the peers exchange their respective URLs, authentication tokens or TLS certificates, ILP addresses, and settlement-related details.
+
+Each ILP Prepare packet is sent as the body of an HTTP request to the peer's server endpoint. ILP Fulfill or Reject packets are returned as the body of the HTTP response.
+
+## Specification
+
+This is a minimal protocol built on HTTP. HTTP/2 is HIGHLY RECOMMENDED for performance reasons, although HTTP/1.1 MAY also be used. Implementations SHOULD support HTTP version negotiation via Application Protocol Negotiation (ALPN).
+
+### Authentication
+
+Peers MAY use any standard HTTP authentication mechanism to authenticate incoming requests. Bearer Tokens or TLS Client Certificates are RECOMMENDED.
+
+### Request
+
+```http
+POST /ilp HTTP/x.x
+Accept: application/ilp+octet-stream
+Content-Type: application/ilp+octet-stream
```

Comment:
I don't think custom content types are necessary unless you plan on doing some kind of protocol negotiation or having other protocols served from the same endpoint.

This is just adding additional implementation burden on top of standard libraries IMO.

On the other hand, if you were sending the packet headers as HTTP headers you could use the types `application/ilp-prepare+octet-stream`, `application/ilp-fulfill+octet-stream`, and `application/ilp-reject+octet-stream` to indicate the type (possibly useful in the load balancer processing too to have this surfaced in the headers).

Reply:
I don't feel strongly about this. @sharafian want to weigh in?

Reply:
I don't think we should be putting the internals of the packet into the headers. I guess custom content types may not be used but I can imagine some cases where you might want to use them (i.e. a middleware that catches all ILP packets but doesn't require a special pathname, just like we do for SPSP).

Reply:
> ...unless you plan on doing some kind of protocol negotiation or having other protocols served from the same endpoint.

I prefer using custom content-types to make room for ideas like the quote cited above...for example, I plan to build both "packet serialization" variants of this proposal (HTTP headers vs OER Packet in the body) to better understand performance, and having a narrower content-type would be helpful here to allow room for other types (e.g., `application/ilp-header+octet-stream`).

Reply:
I'm eager to resolve this discussion because ultimately it doesn't make a big difference whether we use a custom content type or not and we could bikeshed this topic indefinitely.

If we want to use a custom type in the future for a different sort of protocol (maybe along the lines of the ILP packet headers in the HTTP headers), that would be enough to differentiate the OER-serialized version from the later version.

Does anyone feel _very strongly_ that it should be `application/ilp+octet-stream`?

Reply:
FWIW, I don't feel _very_ strongly. My implementation supports all three content-types.

;)

Reply:
@sappenin that's a good reason to use custom content types and actually means we could allow both serializations (OER in body AND packet headers in HTTP headers) because the `Accept` and `Content-Type` headers will tell the peer what is supported.

Can I suggest the following types:
 - `application/ilp-prepare+octet-stream`
 - `application/ilp-fulfill+octet-stream`
 - `application/ilp-reject+octet-stream`

 - `application/ilp+octet-stream` (where packet is OER encoded in body, therefor type is determined from type indicator byte in packet)

Given that most HTTP modules already have easy ways to bind application logic content types tit seems like supporting both serialisations would be pretty trivial. @sappenin, is that your experience?

Reply:
@adrianhopebailie I agree with your type suggestions in principle, and also like the point by @emschwartz about keeping things simple. 

To that end, let's discuss the details of the header-variant of this protocol in a different PR (I don't see anything in the current PR that would block any of that).